### ISSUE TITLE
The service graph supports nodes of the message queue middleware type

### ISSUE THREAD

**Issue by fyuan1316** [grafana/tempo#3348]
**Is your feature request related to a problem? Please describe.**

In the current service graph, I find that the representation of relationships when making asynchronous service calls via message middleware is not intuitive, and I would prefer to be able to visualize the message middleware nodes.

**Describe the solution you'd like**

In the scenario where services are produced and consumed through the message queue middleware, the final graph only shows the invocation relationship between the producer and the consumer. It hides the message middleware node. This is intuitively no different from the use case where the client requests the server directly.

I think there may be the following benefits if we show message middleware type nodes:
- The topology of the diagram is closer to reality and the original asynchronous call relationship is shown more clearly. The producer service is only responsible for writing data to the message middleware and the consumer is only responsible for reading messages from the message middleware.
- In the edge scenario, if the message is not consumed within the timeout period, the current indicator is discarded.  If we use a similar approach for DB requests, this call information from the producer service to the message middleware is preserved.

In conclusion, I would suggest to display the message middleware directly on the service graph, and as far as the technical implementation is concerned, using a similar approach to the way current DB requests are handled seems to help us achieve this. 

What do you guys think?

---
(To give an example of rabbitmq, below is a graph of the current state vs. the state after the requirements are implemented)

## Now 
![image](https://github.com/grafana/tempo/assets/42265657/31e80587-c02d-4911-833e-2d934defde5e)

## Then
![image](https://github.com/grafana/tempo/assets/42265657/da0b30c9-dc4d-4b46-9e70-4941699296fa)

**Describe alternatives you've considered**

**Additional context**

**Comment by joe-elliott** [comment]
Currently I believe we add the database name to the `server` label and set the `connection_type` label equal to `database`. In your example we set the `client`/`server` names to the producer/consumer and set the connection type to `messaging_system`.  

Do you think we should split the metric in two? Add a new label?

**Comment by fyuan1316** [comment]
@joe-elliott 
Thanks for the quick reply!

>Currently I believe we add the database name to the server label and set the connection_type label equal to database. In your example we set the client/server names to the producer/consumer and set the connection type to messaging_system.

Yes, my idea was to split the raw metrics in two, but with the server name taken from messaging.system (rabbitmq in this case) to represent producer->rabbitmq and consumer->rabbitmq. 

Now
|client | server| connection_type|
|-|-|-|
|otel-demo-consumer|otel-demo-provider| messaging_system |

Then
|client | server| connection_type|desc|
|-|-|-|-|
|otel-demo-consumer|rabbitmq| messaging_system | consumer reads message (Considering that this is a control flow and not a data flow, we should set the consumer to be the client)|
|otel-demo-provider|rabbitmq| messaging_system | producer writes message|

**Comment by joe-elliott** [comment]
I like this idea a lot. 

@grafana/observability-traces-and-profiling will the service graph have any issues with the proposed change above? is there any special logic using the `connection_type=messaging_system` we need to be aware of before making this change?

**Comment by aocenas** [comment]
So it does not seem we use `connection_type`  in any way (though maybe it would be nice to use it in some way). So if you have 2 metrics instead of one we will create 1 more node in between. We basically just treat each combination of `client-server` as an edge in the graph, creating nodes in the ends and that is it.

**Comment by joe-elliott** [comment]
Thx @aocenas!

In that base we can split the metric to create a queue node. @fyuan1316 how do you think we should split the duration histogram?

**Comment by fyuan1316** [comment]
Thanks  @joe-elliott ! 

I'm glad you like the proposal and are helping to confirm and move it forward. Yes, I tend to split it in two to represent the latency of production and consumption.

**Comment by fyuan1316** [comment]
Sorry, I may have misunderstood. According to the database implementation, it is, and it looks like it sets ServerLatencySec to the same value as ClientLatencySec. 
Do you think this is ok?

**Comment by fyuan1316** [comment]
Thanks for your reply!
I agree with your approach. It's a great idea to use existing information to render the middle node of the asynchronous call relationship. I also noticed that another proposal (#3453) provides a more complete solution for asynchronous latency in messaging systems. The handling of metrics might be better placed there.

**Comment by joe-elliott** [comment]
Sounds good! Would you like to create a Grafana issue and reference this one? I would be happy to chime in with support.

**Comment by fyuan1316** [comment]
Great! I have created a Grafana issue and referenced this one. Thank you for your support!

**Comment by kvrhdn** [comment]
Hi, just wanted to share some additional details / make sure we are aligned.

When we track edges between services we try to determine the connection type based upon the span kind and present attributes. This is added as label `connection_type` to the metric. [Possible values](https://github.com/grafana/tempo/blob/main/modules/generator/processor/servicegraphs/store/edge.go#L7-L12) are empty (the default), `messaging_system`, `database` and `virtual_node`.

As an example, in the following topology service `A` sends a request to `B` through a messaging system `Queue`:

```
A ------> [ Queue ] ------> B
```

Tempo will create metrics with the labels:
- `client="A"`
- `server="B"`
- `connection_type="messaging_system"`

_(We don't have any labels that contain the name or ID of the messaging system. In hindsight this seems like a huge oversight.)_


Luckily we already capture the latencies `(1)` and `(2)` in separate metrics:

```
A ------> [ Queue ] ------> B
^---(1)---^       ^---(2)---^
```

1) `traces_service_graph_request_client_seconds` = latency of requests between `A` and `Queue`
2) `traces_service_graph_request_server_seconds` = latency of requests between `Queue` and `B`

As an aside, https://github.com/grafana/tempo/pull/3453 is introducing a new metric `traces_service_graph_request_messaging_system_seconds` that measures the processing latency within the messaging system `(3)`:

```
A ------> [ Queue ] ------> B
          ^--(3)--^
```

#### Current situation

The best Grafana can do with the current data is to mark a connection between two nodes as an async messaging system. When rendering the nodes it would check for `connection_type`, if it's `messaging_system` replace the `-->` for something with a block on it indicating there is a messaging system in between. 

```
A -----> B                becomes                A ------> [ <Messaging system> ] ------> B
```


It can separately show the latencies between each node and the messaging system but the count will be shared.

#### Going forward: adding the name

As @fyuan1316 mentioned here (https://github.com/grafana/tempo/issues/3348#issuecomment-1916999679) we really should include the _name_ of the messaging system in the node graph. We can grab this from the attribute `messaging.system`([part of the semconv](https://opentelemetry.io/docs/specs/semconv/messaging/messaging-spans/#messaging-attributes)).

I'd suggest adding a label to the existing metrics `messaging_system="<value of messaging.system>"`. Grafana can use this value (if present) and use it inside the box.

This is backwards compatible: if the value is missing Grafana can default to a placeholder name.

**Comment by fyuan1316** [comment]
Based on the current implementation, I agree with @kvrhdn that is the best Grafana could do. Adding a label representing the specific implementation of `messaging_system` would be better. What do you think? @joe-elliott

**Comment by aocenas** [comment]
I agree with @kvrhdn that marking edges in some way to show the request went through some `messaging_system` makes sense to me. Just want to make sure this is the final version though. I am not sure if the `Going forward: adding the name` section should aim at actually getting to a result that the issue originally requested.

I feel the original request is doable with changes but not sure if it's actually even beneficial. If I understand it correctly it would do something like this:
![tempo_queue](https://github.com/grafana/tempo/assets/1014802/b0f6b912-e85f-42c7-8efa-12efa996d400)

But the problem IMHO is that we have to merge edges and we are losing some granularity of the data. ie `C1-S1` and `C1-S2` will have to be aggregated into `C1-queue` and then on the other side `C1-S1` and `C2-S1` would have to be aggregated into `queue-S1`. I feel like such a transformation would not actually help better understand the latencies or errors in the system as we would lose the granularity of the data per `client-server` pair.

**Comment by joe-elliott** [comment]
Thank you for a detailed description of the changes in Otel and how it will impact Tempo. 🙏  I agree that we do not have the capacity to map service graph connections through span links. Unfortunately, adding this using the current patterns would be quite difficult.

> Also, the current way of service graph metrics only works when the message is handled pretty soon (within the expiry timeframe).

Yes. This is configurable to allow for longer timeframes which obviously comes with a memory cost. We do this to correctly identify parent/child relationships and create the appropriate metrics for the service graph. Tempo is designed to work with any tracing data and not just OTel which is why we don't use attributes like `peer.service`.

> The virtual node can be solution for both situations, the span links and the 'late' handling.

I like this idea. Obviously it would rely on carefully instrumented services, but would allow for a more accurate service graph. I almost wonder if we want a new service graph processor that relies entirely on correct OTel instrumentation. The old one would work in a wider range of cases but would produce less detailed metrics. The new one could be used if you were confident in 100% OTel coverage and provide a more detailed view.

> I have done a small change locally in Tempo:

I think this change requires some thought, but I like the spirit of it. Do you want to open a PR for discussion? or are you simply raising the issue. Either way we appreciate the help.
